File: 2016/05_HowAboutaNiceGameofChess/doorpass.ts
// ======================================================================
// How About a Nice Game of Chess
//   Advent of Code 2016 Day 05 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           d o o r p a s s . t s
//
// A solver for the Advent of Code 2016 Day 05 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------
import { Md5 } from 'ts-md5/dist/md5';

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                               Doorpass
// ======================================================================

export class Doorpass {
  // Object for How About a Nice Game of Chess
  text: string[];

  part2: boolean;

  id: string;

  constructor(text: string[], part2 = false) {
    // Create a Doorpass object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.id = '';

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.id = this.text[0].trim();
    }
  }

  passwordFromID(verbose = false, limit = 0): string {
    if (verbose) console.log(`passwordFromID: ${limit}`);
    // 1. Start with nothing
    let result = '';
    let index = 0;
    // 2. Loop until we compute the eight character password
    while (result.length < 8 && (limit === 0 || index < limit)) {
      // 3. Check if the hash of the id and the index starts with five zeroes
      const idhash = Md5.hashStr(this.id + index).toString();
      if (idhash.startsWith('00000')) {
        result += idhash[5];
        if (verbose) console.log(`${this.id} ${index} ${result}`);
      }
      index += 1;
    }
    return result;
  }

  passwordFromIDTwo(verbose = false, limit = 0): string {
    if (verbose) console.log(`passwordFromIDTwo: ${limit}`);
    // 1. Start with nothing
    let result = '________';
    let index = 0;
    // 2. Loop until we compute the eight character password
    while (result.indexOf('_') !== -1 && (limit === 0 || index < limit)) {
      // 3. Check if the hash of the id and the index starts with five zeroes
      const idhash = Md5.hashStr(this.id + index).toString();
      if (idhash.startsWith('00000')) {
        const offset = parseInt(idhash[5], 16);
        const letter = idhash[6];
        if (offset < 8 && result[offset] === '_') {
          const letters = result.split('');
          letters[offset] = letter;
          result = letters.join('');
        }
        if (verbose) console.log(`${this.id} ${index} ${result} ${offset} ${letter}`);
      }
      index += 1;
    }
    return result;
  }

  solution(verbose = false, limit = 0): string {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      return '';
    }
    return '';
  }

  partOne(verbose = false, limit = 0): string {
    // Returns the solution for part one

    return this.passwordFromID(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): string {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.passwordFromIDTwo(verbose, limit);
  }
}

// ======================================================================
// end                      d o o r p a s s . t s                     end
// ======================================================================

File: 2016/05_HowAboutaNiceGameofChess/doorpass.py
# ======================================================================
# How About a Nice Game of Chess
#   Advent of Code 2016 Day 05 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         d o o r p a s s . p y
# ======================================================================
"A solver for the Advent of Code 2016 Day 05 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import hashlib

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
ZEROES = "00000"
INDEXES = '01234567'

# ======================================================================
#                                                               Doorpass
# ======================================================================


class Doorpass(object):   # pylint: disable=R0902, R0205
    "Object for How About a Nice Game of Chess"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.doorID = ''

        # 2. Set the door ID
        if self.text and len(self.text) > 0:
            self.doorID = self.text[0]

    def password_one(self):
        "Return the password for part one"

        # 1. Start with nothing
        result = []
        base = 0

        # 2. Loop until we get all 8 characters
        while len(result) < 8:

            # 3. Loop until we get a hash that starts with zeroes
            hstr = 'not zeroes'
            while not hstr.startswith(ZEROES):
                hstr = hashlib.md5((self.doorID + str(base)).encode('utf-8')).hexdigest()
                base += 1

            #  4. The next password digit is the 6th char of the hash
            result.append(hstr[5])

        # 9. Return the password
        return ''.join(result)

    def password_two(self):
        "Return the password for part one"

        # 1. Start with nothing
        result = ['_', '_', '_', '_', '_', '_', '_', '_']
        base = 0

        # 2. Loop until we get all 8 characters
        while '_' in result:

            # 3. Loop until we get a hash that starts with zeroes
            hstr = 'not zeroes'
            while not hstr.startswith(ZEROES):
                hstr = hashlib.md5((self.doorID + str(base)).encode('utf-8')).hexdigest()
                base += 1

            #  4. 6th char (if 0-7) says where to place the 7th character
            char6 = hstr[5]
            if char6 in INDEXES:
                indx = int(char6)
                if result[indx] == '_':
                    result[indx] = hstr[6]
                    print(''.join(result))

        # 9. Return the password
        return ''.join(result)

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.password_one()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.password_two()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      d o o r p a s s . p y                     end
# ======================================================================
